apis/python/src/tiledbsoma/_tdb_handles.py

```diff
@@ -49,7 +53,10 @@ def open(
     if not obj_type:
         raise DoesNotExistError(f"{uri!r} does not exist")
     if obj_type == "array":
-        return ArrayWrapper.open(uri, mode, context, timestamp)
+        if mode == "r" and clib.SOMADataFrame.exists(uri):
+            return DataFrameWrapper.open(uri, mode, context, timestamp)
+        else:
+            return ArrayWrapper.open(uri, mode, context, timestamp)
```

Comment:
As I’ve been reading this, I wonder if it would be possible for these C++-based wrappers to take in the TileDB Array object. This would mean that in this code we don’t have to repeatedly open and close these objects for the `.exists` calls—as it stands, every time we do an open, it will end up falling through every case, so it will look like:

 1. Figure out what type to open:

     1. `SOMADataFrame.exists` check
         1. Open array.
         3. Read metadata, compare with `SOMADataFrame`.
         4. Close array. (Does this one actually involve an RPC though?)
     1. `SOMASomeOtherType.exists` check
         1. Open array.
         2. Read metadata, compare with `SOMASomeOtherType`.
         3. Close array.
    
    and only after some number of those are done, do we actually open the array for reals:

 2. Open array *again*, and use that handle this time.

Even if this wasn’t improved to the "just one open" case, it would probably be best to make it so that we only open twice at most: once to figure out its type, then again to use it.

I would say **this** is the largest concern I have about anything in this change.

Reply:
Thank you for your feedback on this. You're right that calling a series of `SOMAObject.exists` is going to unnecessarily open and close multiple arrays and groups when it could probably be avoided with another design. I'll come up with something else.

Reply:
I've made changes to this design so that this should (eventually) only open the TileDB object once instead of using `exists`. 

In C++, I've added a static method `unique_ptr<SOMAObject> open(std::string)` which is a factory that returns a `unique_ptr` to a `SOMAObject` class (ie. `unique_ptr<SOMADataFrame>`). It calls `tiledb::Object` to check whether the URI is a TileDB array or group. Then opens either a `SOMAArray` or `SOMAGroup`. It reads the metadata to see the actual SOMA type. We then pass the already opened `SOMAArray` or `SOMAGroup` to the concrete SOMA class (eg. `make_unique<SOMADataFrame>(open_soma_array)`.

In Pybind11, we call `SOMAObject::open` and can use `soma_object.type()` to get the type of the SOMA object. We can then cast that to the correct Python binding (eg. `py::cast(dynamic_cast<SOMADataFrame&>(*soma_object)))`).